docs/Grammar.md

```diff
@@ -0,0 +1,155 @@
+# EDL Grammar
+
+This document describes the grammar for the EDL language in [Backus-Naur Form](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form).
+
+Though it is based on the Intel SDK's Ocaml Yacc input file
+[Parser.mly](https://github.com/intel/linux-sgx/blob/master/sdk/edger8r/linux/Parser.mly),
+it more accurrately mirrors oedger8r's hand-written EDL Parser
+([parser.cpp](https://github.com/openenclave/oeedger8r-cpp/blob/master/parser.cpp)).
+
+# Conventions
+
+Keywords in EDL are enclosed in double quotes.
+E.g: `"enclave" "import"`.
+
+Operators in EDL are are represented in single quotes.
+E.g: `'{' ',' '*'`.
+
+Comments and white-spaces are assumed to be removed by the lexer.
+
+In addition to keywords and operators, the lexer recognizes the following tokens
+- `identifier` A sequence of alphanumeric characters or '_' that starts with a character or '_'.
+   Valid identifiers in EDL are valid C language identifiers.
+- `integer` A sequence of digits. EDL does not support octal and hexadecimal integers.
+- `string` A C-style string literal.
+
+Each `non-termnial` is defined via `production` of the form
+```c
+     non-termnial = expression
+```
+where `expression` consists of one or more `non-terminal`s or tokens.
+
+
+`+` Indicates one or more occurence of the preceding symbol.
+
+`*` Indicates zero or more occurence of the preceding symbol.
+
+`[]` Indicates that the enclosed item is optional.
+
+`()` Paranthesis is used to group items.
+
+`|` Indicates a choice between different productions.
+
+## Grammar
+
+```c
+
+enclave_definition :  "enclave" '{' enclave_item* '}' EOF
+
+enclave_item :
+    include_statement
+    | import_statement
+    | from_import_statement
+    | enum_declaration
+    | struct_declaration
+    | union_declaration
+    | trusted_section
+    | untrusted_section
+
+include_statement: "include" string
+
+import_statement: "import" string
+
+from_import_statement: "from" string "import" imported_items ';'
+
+imported_items =
+    '*'
+    | identifier_list
+
+identifier_list = identifier ( ',' identifier )*
+
+enum_declaration =
+    "enum" [identifier] '{' enumeration_item ( ',' enumeration_item )* '}' ';'
+
+enumeration_item =
+    | identifier = integer | identifier
+    | identifier
+
+struct_declaration = "struct" '{' ( declaration ';' ) * '}' ';'
+
+union_declaration = "union" '{' ( declaration ';' ) * '}' ';'
+
+declaration = [attributes] type_expr identifier [array_dimension+]
+
+attributes = '[' attribute ( ',' attribute )* ']'
+
+attribute =
+    "in" | "out" | "string" | "wstring" | "isptr" | "isary" | "user_check"
+    | "count" = attribute_value
+    | "size" = attribute_value
```

Comment:
can we differentiate the use of `count` and `size`? do you think we can add something here?

Reply:
Good point. Introduced a `size_or_count` non-terminal. Also the = needs to be `=`